runtime/lua/vim/lsp/sync.lua

```diff
@@ -93,30 +93,35 @@ end
 -- utf-8 index and either the utf-16, or utf-32 index.
 ---@param line string the line to index into
 ---@param byte integer the byte idx
+---@param start boolean true for start align, false for end align
 ---@param offset_encoding string utf-8|utf-16|utf-32|nil (default: utf-8)
 ---@returns table<string, int> byte_idx and char_idx of first change position
-local function align_end_position(line, byte, offset_encoding)
+local function align_position(line, byte, start, offset_encoding)
+  if byte <= #line then
+    -- Modifying line, find the nearest utf codepoint
+    local offset = str_utf_start(line, byte)
+
+    -- If the byte does not fall on the start of the character, then
+    -- align to the start of the next character if end align, and start
+    -- of this character if start align
+    if offset < 0 then
+      if start then
+        byte = byte + offset
+      else
+        byte = byte + str_utf_end(line, byte) + 1
+      end
+    end
+  end
+
   local char
-  -- If on the first byte, or an empty string: the trivial case
-  if byte == 1 or #line == 0 then
```

Comment:
Why did you remove this optimization?

Reply:
My bad, just added an equivalent optimization for this case?

Reply:
This isn't an equivalent optimization. It doesn't short-ircuit the valuation as it did previously. You're still checking the byte offset.

Reply:
Good catch, just fixed.

Reply:
Hmm is there not a scenario in which byte > #line and #line == 0 that this is not catching?

Reply:
In this case `char` should be 1 right?

Reply:
In the current implementation, that is the case

Reply:
It should be the case here as well. In `compute_end_range()`, if one of the lines has length 0 the `byte_offset` stays at -1. (`compute_start_range()` wasn't really changed)